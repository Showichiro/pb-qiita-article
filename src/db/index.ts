export * as schema from "./schema";
export * from "./findAllArticles";
export * from "./getArticleCountGroupByUser";
export * from "./getLikesCountGroupByUser";
