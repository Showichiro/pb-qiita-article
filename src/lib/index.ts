export * from "./drizzle";
export * from "./hono";
export * from "./zod";
