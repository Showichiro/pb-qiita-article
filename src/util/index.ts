export * from "./typeUtils";
export * from "./testUtils";
export * from "./dateFormatUtils";
export * from "./factory";
export * from "./dateParamsUtils";
