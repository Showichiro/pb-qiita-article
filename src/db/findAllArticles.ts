import { DrizzleD1Database } from "@/lib";
import * as schema from "./schema";
import { Article } from "@/schemas";

/**
 * Find all articles.
 *
 * @param db - Database instance.
 * @param config - Configuration object.
 * @returns Array of articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
type FindAllArticlesReturnType = Array<Article>;

/**
 * Configuration object.
 *
 * @param limit - Limit number of articles.
 * @param offset - Offset number of articles.
 * @param sinse - Date since articles.
 * @param until - Date until articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
export type FindAllArticlesConfig = {
  limit: number | null;
  offset: number | null;
  sinse: string | null;
  until: string | null;
};

/**
 * Find all articles.
 *
 * @param db - Database instance.
 * @param config - Configuration object.
 * @returns Array of articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
export const findAllArticles = async (
  db: DrizzleD1Database<typeof schema>,
  config: FindAllArticlesConfig
): Promise<FindAllArticlesReturnType> => {
  const defaultLimit = 10;
  const defaultOffset = 0;
  const { limit, offset, sinse, until } = config;
  const results = await db.query.articles.findMany({
    limit: limit ?? defaultLimit,
    offset: offset ?? defaultOffset,
    where:
      !sinse && !until
        ? undefined
        : (fileds, { between, gte, lte }) => {
            if (sinse && until) {
              return between(fileds.createdAt, sinse, until);
            }
            if (sinse) {
              return gte(fileds.createdAt, sinse);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          },
    orderBy: (fileds, { desc }) => [desc(fileds.createdAt)],
    with: {
      tags: {
        columns: {
          articleId: false,
          id: false,
          name: true,
        },
      },
    },
  });
  return results;
};
