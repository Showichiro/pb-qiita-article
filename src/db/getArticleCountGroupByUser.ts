import { DrizzleD1Database, asc, count, desc, between, gte, lte } from "@/lib";
import { schema } from "@/db";
import { ArticleCountGroupByUser } from "@/schemas";

/**
 * Config for getArticleCountGroupByUser.
 * @property {string | null} sinse Start date.
 * @property {string | null} until End date.
 * @property {"asc" | "desc"} sort Sort order. Default is "desc".
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: "2024-01-01T00:00:00Z",
 *   until: "2023-01-01T00:00:00Z",
 *   sort: "asc",
 * };
 * ```
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: null,
 *   until: null,
 *   sort: "asc",
 * };
 * ```
 */
export type RankingConfig = {
  sinse: string | null;
  until: string | null;
  sort?: "asc" | "desc";
};

/**
 * Get article count group by user.
 * @param db DrizzleD1Database instance.
 * @param config Config for getArticleCountGroupByUser.
 * @returns Array of ArticleCountGroupByUser.
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: "2024-01-01T00:00:00Z",
 *   until: "2023-01-01T00:00:00Z",
 *   sort: "asc",
 * };
 * const results = await getArticleCountGroupByUser(db, config);
 * ```
 */
export const getArticleCountGroupByUser = async (
  db: DrizzleD1Database<typeof schema>,
  { sinse, until, sort = "desc" }: RankingConfig
): Promise<Array<ArticleCountGroupByUser>> => {
  const results = await db
    .select({
      count: count(schema.articles.id),
      userId: schema.articles.userId,
      userName: schema.articles.userName,
      createdAt: schema.articles.createdAt,
    })
    .from(schema.articles)
    .where(
      !sinse && !until
        ? undefined
        : (fileds) => {
            if (sinse && until) {
              return between(fileds.createdAt, sinse, until);
            }
            if (sinse) {
              return gte(fileds.createdAt, sinse);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          }
    )
    .groupBy((fileds) => fileds.userId)
    .orderBy((fileds) => [
      sort === "asc" ? asc(fileds.count) : desc(fileds.count),
    ])
    .all();
  return results.map(({ createdAt, ...rest }) => ({ ...rest }));
};
