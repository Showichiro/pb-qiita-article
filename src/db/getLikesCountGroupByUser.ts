import { asc, between, desc, type DrizzleD1Database, gte, lte, sum } from "@/lib";
import type { LikesCountSchema } from "@/schemas";
import { type RankingConfig, schema } from "@/db";

/**
 * Get likes count group by user
 * @param db
 * @param config
 * @returns Array<LikesCountSchema>
 * @example
 * getLikesCountGroupByUser(db, {
 *   since: "2024-01-01T00:00:00Z",
 *   until: null,
 *   sort: "asc",
 * });
 */
export const getLikesCountGroupByUser = async (
  db: DrizzleD1Database<typeof schema>,
  { since, until, sort = "desc" }: RankingConfig,
): Promise<Array<LikesCountSchema>> => {
  const results = await db
    .select({
      totalLikesCount: sum(schema.articles.likesCount),
      userId: schema.articles.userId,
      userName: schema.articles.userName,
      createdAt: schema.articles.createdAt,
    })
    .from(schema.articles)
    .where(
      !since && !until
        ? undefined
        : (fileds) => {
            if (since && until) {
              return between(fileds.createdAt, since, until);
            }
            if (since) {
              return gte(fileds.createdAt, since);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          },
    )
    .groupBy((fileds) => fileds.userId)
    .orderBy((fileds) => [
      sort === "asc"
        ? asc(fileds.totalLikesCount)
        : desc(fileds.totalLikesCount),
    ])
    .all();
  return results.map(({ createdAt, ...rest }) => ({
    ...rest,
  }));
};
