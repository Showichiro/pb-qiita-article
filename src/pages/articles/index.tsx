import {
  Header,
  PageLayout,
  Pagenation,
  PageTitle,
  SelectBoxPageLimit,
  Spinner,
} from "@/components";
import { ArticlesContainer } from "@/containers";
import { FindAllArticlesConfig, schema } from "@/db";
import { DrizzleD1Database } from "@/lib";
import { FC, Suspense } from "hono/jsx";

export const ArticlesPage: FC<{
  db: DrizzleD1Database<typeof schema>;
  config: FindAllArticlesConfig;
}> = ({ config, db }) => {
  const limit = config.limit ?? 10;
  const page = (config.offset ?? 0) / limit + 1;
  return (
    <>
      <Header />
      <PageTitle label="記事一覧" />
      <PageLayout>
        <Suspense fallback={<Spinner />}>
          <ArticlesContainer db={db} config={config} />
        </Suspense>
        <div class="my-4 flex">
          <div>
            <Pagenation page={page} limit={limit} />
          </div>
          <div class="ml-2">
            <SelectBoxPageLimit page={page} limit={limit} />
          </div>
        </div>
      </PageLayout>
    </>
  );
};
