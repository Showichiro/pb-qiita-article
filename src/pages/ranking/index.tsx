import { RankingRange } from "@/components";
import { RankningContainer } from "@/containers";
import { RankingConfig, schema } from "@/db";
import { DrizzleD1Database } from "@/lib";
import { dateTimetoDateString } from "@/util/dateFormatUtils";
import { FC, Suspense } from "hono/jsx";

export const RankingPage: FC<{
  db: DrizzleD1Database<typeof schema>;
  config: RankingConfig;
}> = ({ config, db }) => {
  return (
    <>
      <h1>記事数・いいね数ランキング</h1>
      <RankingRange
        default={{
          since: config.since ? dateTimetoDateString(config.since) : null,
          until: config.until ? dateTimetoDateString(config.until) : null,
        }}
      />
      <Suspense fallback={"...loading"}>
        <RankningContainer config={config} db={db} />
      </Suspense>
    </>
  );
};
