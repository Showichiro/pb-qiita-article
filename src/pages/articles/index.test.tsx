import { schema } from "@/db";
import { drizzle } from "@/lib";
import { renderer } from "@/util";
import { Miniflare } from "miniflare";
import { ArticlesPage } from "@/pages";

describe("ArticlesPage", async () => {
  const mf = new Miniflare({
    modules: true,
    d1Databases: ["DB"],
    script: `addEventListener("fetch", (event) => {
              event.respondWith(new Response("Hello Miniflare!"));
            })`,
  });

  const db = await mf.getD1Database("DB");

  const record = 10;
  beforeAll(async () => {
    await db.exec(
      "CREATE TABLE `articles` (`id` text PRIMARY KEY NOT NULL,`title` text NOT NULL,`user_id` text NOT NULL,`user_name` text NOT NULL,`created_at` text NOT NULL,`likes_count` integer NOT NULL,`stocks_count` integer NOT NULL);",
    );
    await db.exec(
      "CREATE TABLE `tags` (`article_id` text,`id` integer PRIMARY KEY NOT NULL,`name` text NOT NULL,FOREIGN KEY (`article_id`) REFERENCES `articles`(`id`) ON UPDATE cascade ON DELETE cascade);",
    );
    const promises = [...Array(record)].map(async (_, index) => {
      return await db
        .prepare(
          "INSERT INTO `articles` (`id`, `title`, `user_id`, `user_name`, `created_at`, `likes_count`, `stocks_count`) VALUES (?, ?, ?, ?, ?, ?, ?);",
        )
        .bind(
          `${index}`,
          `title-${index}`,
          `user-${index % 2 === 0 ? 0 : index}`,
          `user-${index % 2 === 0 ? 0 : index}`,
          new Date(index).toISOString(),
          0,
          0,
        )
        .run()
        .then(async () => {
          return await db
            .prepare("INSERT INTO `tags` (`article_id`, `name`) VALUES (?, ?);")
            .bind(`${index}`, `tag-${index}`)
            .run();
        });
    });
    await Promise.all(promises);
  });

  afterAll(async () => {
    await mf.dispose();
  });

  const instance = drizzle(db, { schema, logger: true });

  it("should render article page", async () => {
    const { text } = await renderer(
      <ArticlesPage
        config={{ limit: null, offset: null, since: null, until: null }}
        db={instance}
      />,
    );
    expect(text).toMatchSnapshot();
  });
});
