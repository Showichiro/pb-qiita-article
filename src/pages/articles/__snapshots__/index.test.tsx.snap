// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticlesPage > should render article page 1`] = `
"<div class="navbar bg-base-100"><div class="navbar-start"><div class="dropdown"><div tabindex="0" role="button" aria-label="navbar-button" class="btn btn-ghost btn-circle"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h7"></path></svg></div><ul tabindex="0" class="menu menu-sm dropdown-content mt-3 z-[1] p-2 shadow bg-base-100 rounded-box w-52"><li><a href="/articles">記事一覧</a></li><li><a href="/ranking">いいね・投稿数ランキング</a></li></ul></div></div><div class="navbar-center"><a class="btn btn-ghost text-xl" href="https://qiita.com/organizations/primebrains" target="_blank" rel="noopener noreferrer">PB Qiita</a></div><div class="navbar-end"><ul class="menu menu-horizontal px-1"></ul></div></div><h1 class="text-4xl m-4">記事一覧</h1><div class="mt-10 mx-4"><table class="table overflow-x-auto"><thead><tr><th>タイトル</th><th>執筆者</th><th>タグ</th><th>いいね数</th><th>ストック数</th><th>投稿日</th></tr></thead><tbody><tr class="hover"><td><a class="link" href="https://qiita.com/user-9/items/9" target="_blank" rel="noopener noreferrer">title-9</a></td><td><a class="link" href="https://qiita.com/user-9" target="_blank" rel="noopener noreferrer">user-9<span>(user-9)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-9" target="_blank" rel="noopener noreferrer">tag-9</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/8" target="_blank" rel="noopener noreferrer">title-8</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-8" target="_blank" rel="noopener noreferrer">tag-8</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-7/items/7" target="_blank" rel="noopener noreferrer">title-7</a></td><td><a class="link" href="https://qiita.com/user-7" target="_blank" rel="noopener noreferrer">user-7<span>(user-7)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-7" target="_blank" rel="noopener noreferrer">tag-7</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/6" target="_blank" rel="noopener noreferrer">title-6</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-6" target="_blank" rel="noopener noreferrer">tag-6</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-5/items/5" target="_blank" rel="noopener noreferrer">title-5</a></td><td><a class="link" href="https://qiita.com/user-5" target="_blank" rel="noopener noreferrer">user-5<span>(user-5)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-5" target="_blank" rel="noopener noreferrer">tag-5</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/4" target="_blank" rel="noopener noreferrer">title-4</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-4" target="_blank" rel="noopener noreferrer">tag-4</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-3/items/3" target="_blank" rel="noopener noreferrer">title-3</a></td><td><a class="link" href="https://qiita.com/user-3" target="_blank" rel="noopener noreferrer">user-3<span>(user-3)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-3" target="_blank" rel="noopener noreferrer">tag-3</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/2" target="_blank" rel="noopener noreferrer">title-2</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-2" target="_blank" rel="noopener noreferrer">tag-2</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-1/items/1" target="_blank" rel="noopener noreferrer">title-1</a></td><td><a class="link" href="https://qiita.com/user-1" target="_blank" rel="noopener noreferrer">user-1<span>(user-1)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-1" target="_blank" rel="noopener noreferrer">tag-1</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/0" target="_blank" rel="noopener noreferrer">title-0</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-0" target="_blank" rel="noopener noreferrer">tag-0</a></li></ul></td><td>0</td><td>0</td><td>1970-01-01</td></tr></tbody></table><div class="my-4 flex"><div><div class="join"><form action="/articles" method="get"><input type="hidden" name="limit" value="10"/><button type="submit" name="offset" value="-10" disabled="" class="join-item btn">Prev</button><span class="join-item btn">1</span><button class="join-item btn" type="submit" name="offset" value="10">Next</button></form></div></div><div class="ml-2"><select class="select select-bordered" aria-label="page-limit" id="page-limit" name="page-limit"><option selected="">10</option><option>20</option><option>30</option></select>
    <script>
      const page = 1;
      const handleSelect = (e) => {
        const limit = e.target.value;
        const offset = (page - 1) * limit;
        const url = new URL(location.href);
        url.searchParams.set("limit", limit);
        url.searchParams.set("offset", offset);
        location.href = url.href;
      };
      const select = document.querySelector("select#page-limit");
      select.addEventListener("change", handleSelect);
    </script>
  </div></div></div>"
`;
